server/src/main/java/org/opensearch/cluster/routing/allocation/decider/AwarenessAllocationDecider.java

```diff
@@ -174,53 +174,51 @@ private Decision underCapacity(ShardRouting shardRouting, RoutingNode node, Rout
+            int numShardCopiesAllocatedToCurrentNodeAttributeValue = getNumberOfShardCopiesAllocatedToCurrentNodeAttributeValue(shardRouting, node, allocation, moveToNode, awarenessAttribute);
 
-            // build attr_value -> nodes map
-            Set<String> nodesPerAttribute = allocation.routingNodes().nodesPerAttributesCounts(awarenessAttribute);
-            int numberOfAttributes = nodesPerAttribute.size();
+            Set<String> attributeValues = allocation.routingNodes().nodesPerAttributesCounts(awarenessAttribute);
+            int distinctAttributeValuesCount = attributeValues.size();
             List<String> fullValues = forcedAwarenessAttributes.get(awarenessAttribute);
 
             if (fullValues != null) {
-                // If forced awareness is enabled, numberOfAttributes = count(distinct((union(discovered_attributes, forced_attributes)))
+                // If forced awareness is enabled, distinctAttributeValuesCount = count(distinct((union(discovered_attributes, forced_attributes)))
                 Set<String> attributesSet = new HashSet<>(fullValues);
-                for (String stringObjectCursor : nodesPerAttribute) {
-                    attributesSet.add(stringObjectCursor);
-                }
-                numberOfAttributes = attributesSet.size();
+                attributesSet.addAll(attributeValues);
+                distinctAttributeValuesCount = attributesSet.size();
             }
 
             // TODO should we remove ones that are not part of full list?
-            final int maximumNodeCount = (shardCount + numberOfAttributes - 1) / numberOfAttributes; // ceil(shardCount/numberOfAttributes)
-            if (currentNodeCount > maximumNodeCount) {
+            final int maximumShardCopiesAllowedPerAttributeValue = (shardCopiesCount + distinctAttributeValuesCount - 1) / distinctAttributeValuesCount; // ceil(shardCopiesCount/distinctAttributeValuesCount)
+            if (numShardCopiesAllocatedToCurrentNodeAttributeValue > maximumShardCopiesAllowedPerAttributeValue) {
                 return allocation.decision(
                     Decision.NO,
                     NAME,
-                    "there are too many copies of the shard allocated to nodes with attribute [%s], there are [%d] total configured "
+                    "there are too many copies of the shard allocated to nodes with attribute [%s] and attribute value [%s], there are [%d] total configured "
                         + "shard copies for this shard id and [%d] total attribute values, expected the allocated shard count per "
                         + "attribute [%d] to be less than or equal to the upper bound of the required number of shards per attribute [%d]",
                     awarenessAttribute,
-                    shardCount,
-                    numberOfAttributes,
-                    currentNodeCount,
-                    maximumNodeCount
+                    getAttributeValueForNode(node, awarenessAttribute),
```

Comment:
Can we remove this extra parameter added? HashMap.get call is very expensive operation.